libsql-server/src/connection/connection_core.rs

```diff
@@ -0,0 +1,809 @@
+            block_writes,
+            resolve_attach_path,
+            forced_rollback: false,
+            broadcaster,
+            hooked: false,
+            canceled,
+        };
+
+        for ext in extensions.iter() {
+            unsafe {
+                let _guard = rusqlite::LoadExtensionGuard::new(&this.conn).unwrap();
+                if let Err(e) = this.conn.load_extension(ext, None) {
+                    tracing::error!("failed to load extension: {}", ext.display());
+                    Err(e)?;
+                }
+                tracing::trace!("Loaded extension {}", ext.display());
+            }
+        }
+
+        Ok(this)
+    }
+
+    pub(super) fn raw_mut(&mut self) -> &mut libsql_sys::Connection<W> {
+        &mut self.conn
+    }
+
+    pub(super) fn raw(&self) -> &libsql_sys::Connection<W> {
+        &self.conn
+    }
+
+    pub(super) fn config(&self) -> Arc<DatabaseConfig> {
+        self.config_store.get()
+    }
+
+    pub(super) async fn run_async<B: QueryResultBuilder>(
+        this: Arc<Mutex<Self>>,
+        pgm: Program,
+        builder: B,
+    ) -> Result<B> {
+        struct Bomb {
+            canceled: Arc<AtomicBool>,
+            defused: bool,
+        }
+
+        impl Drop for Bomb {
+            fn drop(&mut self) {
+                if !self.defused {
+                    tracing::trace!("cancelling request");
+                    self.canceled.store(true, Ordering::Relaxed);
+                }
+            }
+        }
+
+        let canceled = {
+            let cancelled = this.lock().canceled.clone();
+            cancelled.store(false, Ordering::Relaxed);
+            cancelled
+        };
+
+        PROGRAM_EXEC_COUNT.increment(1);
+
+        // create the bomb right before spawning the blocking task.
+        let mut bomb = Bomb {
+            canceled,
+            defused: false,
+        };
+        let ret = BLOCKING_RT
+            .spawn_blocking(move || CoreConnection::run(this, pgm, builder))
+            .await
+            .unwrap();
+
+        bomb.defused = true;
+
+        ret
+    }
+
+    pub(super) fn run<B: QueryResultBuilder>(
+        this: Arc<Mutex<Self>>,
+        pgm: Program,
+        mut builder: B,
+    ) -> Result<B> {
+        let (config, stats, block_writes, resolve_attach_path) = {
+            let mut lock = this.lock();
+            let config = lock.config_store.get();
+            let stats = lock.stats.clone();
+            let block_writes = lock.block_writes.clone();
+            let resolve_attach_path = lock.resolve_attach_path.clone();
+
+            lock.update_hooks();
+
+            (config, stats, block_writes, resolve_attach_path)
+        };
+
+        builder.init(&this.lock().builder_config)?;
+        let mut vm = Vm::new(
+            builder,
+            &pgm,
+            move |stmt_kind| {
+                let should_block = match stmt_kind {
+                    StmtKind::Read | StmtKind::TxnBegin => config.block_reads,
```

Comment:
hmmm we don't check for block_reads here like we do for writes?

Reply:
No idea, I just copied the code over here no additional changes